### DISCUSSION TITLE
Effects and Transitions

### DISCUSSION THREAD

**Discussion by guerratron** [frang75/nappgui_src#101] [General]
Greetings everyone, I'm starting to get interested in this project, although I'm a newbie.
I was wondering if **effects** and **transitions** can be implemented.
...I don't know, I've searched and can't find anything about it...

> _Sorry for my English, I use google-translate_

**Comment by frang75** [comment]
Hi @guerratron and welcome to NAppGUI.
I'm not sure what you mean by Effects and Transitions. With NAppGUI you can create Custom Views where you can implement any type of drawings or animations.

**Comment by guerratron** [comment]
Thanks for responding, frang75.

I was looking for transitions between panels, for example to simulate a gradual _fade-in/fade-out_, or perhaps when selecting to show a given column/row with "_layout_show_col(..)_" or "_layout_show_row(..)_".
I thought maybe through the "**Clock**" object I can move controls until they disappear...

Anyway, I'll take a look at what you suggest "**Custom Views**", I assume you mean this:   

```c
View* view_custom(const bool_t scroll, const bool_t border);
```

Thank you so much.

**Comment by guerratron** [comment]
I've been looking into the API and maybe it will allow me to move **GUI** elements and apply effects if I use tasks

```c
osapp_task(tdata, .04, i_task_main, i_task_update, i_task_end, TaskData);
```
I'll try this afternoon to see if I can do it.

**Reply by frang75** [reply]
The best way (and safer) to create animations in GUI widgets or Custom views is using the `osmain_sync` instead `osmain`, which add the `update` method.

![listbox_animation](https://github.com/frang75/nappgui_src/assets/42999199/47bf340f-6720-450c-8a60-407b7c3a7981)

```
/* NAppGUI Hello World */

#include <nappgui.h>

typedef struct _app_t App;

struct _app_t
{
    Window *window;
    TextView *text;
    ListBox *listbox;
    real64_t time;
    uint32_t clicks;
    uint32_t state;
};

/*---------------------------------------------------------------------------*/

static void i_OnButton(App *app, Event *e)
{
    textview_printf(app->text, "Button click (%d)\n", app->clicks);
    app->clicks += 1;
    unref(e);
}

/*---------------------------------------------------------------------------*/

static Panel *i_panel(App *app)
{
    Panel *panel = panel_create();
    Layout *layout = layout_create(1, 4);
    Label *label = label_create();
    Button *button = button_push();
    TextView *text = textview_create();
    ListBox *list = listbox_create();
    app->text = text;
    app->listbox = list;
    label_text(label, "Hello!, I'm a label");
    button_text(button, "Click Me!");
    button_OnClick(button, listener(app, i_OnButton, App));
    listbox_add_elem(list, "Item1", NULL);
    listbox_add_elem(list, "Item2", NULL);
    listbox_add_elem(list, "Item3", NULL);
    listbox_add_elem(list, "Item4", NULL);
    layout_label(layout, label, 0, 0);
    layout_button(layout, button, 0, 1);
    layout_textview(layout, text, 0, 2);
    layout_listbox(layout, list, 0, 3);
    layout_hsize(layout, 0, 250);
    layout_vsize(layout, 2, 100);
    layout_margin(layout, 5);
    layout_vmargin(layout, 0, 5);
    layout_vmargin(layout, 1, 5);
    layout_vmargin(layout, 2, 5);
    panel_layout(panel, layout);
    return panel;
}

/*---------------------------------------------------------------------------*/

static void i_OnClose(App *app, Event *e)
{
    osapp_finish();
    unref(app);
    unref(e);
}

/*---------------------------------------------------------------------------*/

static App *i_create(void)
{
    App *app = heap_new0(App);
    Panel *panel = i_panel(app);
    app->window = window_create(ekWINDOW_STD);
    app->state = 0;
    app->time = 0;
    window_panel(app->window, panel);
    window_title(app->window, "Hello, World!");
    window_origin(app->window, v2df(500, 200));
    window_OnClose(app->window, listener(app, i_OnClose, App));
    window_show(app->window);
    return app;
}

/*---------------------------------------------------------------------------*/

static void i_destroy(App **app)
{
    window_destroy(&(*app)->window);
    heap_delete(app, App);
}

/*---------------------------------------------------------------------------*/

static void i_update(App *app, const real64_t ptime, const real64_t ctime)
{
    if (ctime - app->time > 0.3)
    {
        if (app->state == 0)
            listbox_color(app->listbox, 0, kCOLOR_RED);
        else if (app->state == 1)
            listbox_color(app->listbox, 0, kCOLOR_GREEN);
        else if (app->state == 2)
            listbox_color(app->listbox, 0, kCOLOR_BLUE);

        app->time = ctime;
        app->state += 1;
        if (app->state == 3)
            app->state = 0;
    }
}

/*---------------------------------------------------------------------------*/

#include "osmain.h"
osmain_sync(0.04, i_create, i_destroy, i_update, "", App)
```

**Comment by guerratron** [comment]
Yeah! this is the key.
The "**update**" method. Well, thank you! I'll go that way and if I get something interesting I'll show it here.
Thank you very much, @frang75.
You have done an impressive job with the **SDK**, also with the "_cMakes helpers_". My respects

**Reply by frang75** [reply]
We have also a Discord server. Email me with your username and I'll add you to de server.

**Comment by guerratron** [comment]
thank you very kind.
Discord username: guerratron24

El lun, 12 feb 2024 a las 22:20, Francisco García Collado (<
***@***.***>) escribió:

> We have also a Discord server. Email me with your username and I'll add
> you to de server.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/frang75/nappgui_src/discussions/101#discussioncomment-8446998>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDGJKA44LZAYCWIOP7IONDYTKBRXAVCNFSM6AAAAABDCF6NSWVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DINBWHE4TQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
... que la fuerza t'acompañe ..

**Comment by guerratron** [comment]
Greetings,
Around the *effects-transitions* I wanted to build something like transition **tabs** or similar.
This is what I have managed to do, as if it were a kind of *"widget"* to add content in the form of a "**showcase**" (for example for a visualization of options)..

https://github.com/frang75/nappgui_src/assets/4613288/c5220d77-05d2-44f8-a9ff-d6171bd42201


https://github.com/frang75/nappgui_src/assets/4613288/ebaef97a-a308-442b-944e-d3c2090701dd


..it is very basic and lacks stylization...
